src/ellswift.rs

```diff
@@ -0,0 +1,605 @@
+//!
+//! Then an ElligatorSwift encoding of x consists of the 32-byte big-endian
+//! encodings of field elements u and t concatenated, where f(u,t) = x.
+//! The encoding algorithm is described in the paper, and effectively picks a
+//! uniformly random pair (u,t) among those which encode x.
+//!
+//! If the Y coordinate is relevant, it is given the same parity as t.
+//!
+//! Changes w.r.t. the the paper:
+//! - The u=0, t=0, and u^3+t^2+7=0 conditions result in decoding to the point
+//!   at infinity in the paper. Here they are remapped to finite points.
+//! - The paper uses an additional encoding bit for the parity of y. Here the
+//!   parity of t is used (negating t does not affect the decoded x coordinate,
+//!   so this is possible).
+
+use core::ops::Deref;
+
+use ffi::{secp256k1_ellswift_xdh, EllswiftECDHHashFn};
+
+use crate::ffi::{
+    self, secp256k1_ellswift_create, secp256k1_ellswift_decode, secp256k1_ellswift_encode, CPtr,
+};
+use crate::{constants, from_hex, Context, PublicKey, Secp256k1, SecretKey};
+/// ElligatorSwift is an encoding of a uniformly chosen point on the curve
+/// as a 64-byte array that is indistinguishable from a uniformly random.
+/// This object holds tow field elements u and t, which are the inputs to
+/// the ElligatorSwift encoding function.
+#[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct ElligatorSwift(ffi::ElligatorSwift);
+
+impl ElligatorSwift {
+    /// Create a new ElligatorSwift object from a 64-byte array.
+    pub fn ellswift_create(priv32: SecretKey, rand32: [u8; 32]) -> ElligatorSwift {
+        let mut ell_out = [0u8; constants::ELLSWIFT_ENCODING_SIZE];
+        unsafe {
+            secp256k1_ellswift_create(
+                ffi::secp256k1_context_no_precomp,
+                ell_out.as_mut_c_ptr(),
+                priv32.as_c_ptr(),
+                rand32.as_ptr(),
+            );
+        }
+        ElligatorSwift(ffi::ElligatorSwift::from_array(ell_out))
+    }
+    /// Decode an ElligatorSwift object from a 64-byte array.
+    pub fn from_array(ser: [u8; 64]) -> ElligatorSwift {
+        ElligatorSwift(ffi::ElligatorSwift::from_array(ser))
+    }
+    /// Creates the Elligator Swift encoding from a secret key, using some aux_rand if defined.
+    /// This method is preferred instead of just decoding, because the private key offers extra
+    /// entropy.
+    /// # Example
+    /// ```
+    /// use secp256k1::{ellswift::ElligatorSwift, PublicKey, Secp256k1, SecretKey};
+    /// let secp = Secp256k1::new();
+    /// let sk = SecretKey::from_slice(&[1; 32]).unwrap();
+    /// let es = ElligatorSwift::from_seckey(&secp, sk, None);
+    /// ```
+    pub fn from_seckey<C: Context>(
+        secp: &Secp256k1<C>,
+        sk: SecretKey,
+        aux_rand: Option<[u8; 32]>,
+    ) -> ElligatorSwift {
+        let mut es_out = [0u8; constants::ELLSWIFT_ENCODING_SIZE];
+        let aux_rand = aux_rand.map(|rand| rand.as_c_ptr()).unwrap_or(std::ptr::null());
```

Comment:
aux_rand here is a dangling pointer, you should either make it an `Option<&[u8; 32]>` or do:

`let aux_rand_ptr = aux_rand.as_ref().map(|rand| rand.as_c_ptr()).unwrap_or(std::ptr::null());`

Reply:
Hey @elichai I can't work out your comment, doesn't the array coerce to a slice and `as_ptr` is called on it the same as it would be in your code?

Reply:
Let's look at this function expanded a bit:
```rust
let aux_rand: Option<[u8; 32]; // This is on the function's stack as a function argument
let map_function: FnOnce([u8; 32]) -> *const u8 = {
     // This is a new stack for this function.
    let aux_rand: [u8; 32]; // this is on the closure's stack as a function argument.
    let ptr = <[u8]>::as_c_ptr(deref_to_slice(&aux_rand)); // this creates a slice pointing at `aux_rand` in the closure's stack, it then passes it to `as_c_ptr` which then returns a pointer to that slice.
    ptr // returns a pointer to the `aux_rand` that is on the closure's stack, and will get invalidated immediately 
};
let aux_rand = Option::map(aux_rand, map_function); // Copies the original `aux_rand` into the `Option::map` stack (which will then pass it into `map_function`)

```